File: python/src/metadata_ai/_http.py
"""HTTP client internals for the Metadata AI SDK.

This module provides HTTP clients with:
- Automatic retry with exponential backoff
- Request correlation IDs for debugging
- Proper logging integration
- Error handling with specific exceptions
"""

from __future__ import annotations

import json as json_module
import time
import uuid
from collections.abc import AsyncIterator, Iterator
from typing import Any

import httpx

from metadata_ai._logging import get_logger
from metadata_ai.auth import TokenAuth
from metadata_ai.exceptions import (
    AgentExecutionError,
    AgentNotEnabledError,
    AgentNotFoundError,
    AuthenticationError,
    BotNotFoundError,
    MetadataError,
    RateLimitError,
)

logger = get_logger(__name__)

# Status codes that should trigger retry
RETRYABLE_STATUS_CODES = {429, 500, 502, 503, 504}


def _generate_request_id() -> str:
    """Generate a unique request ID for correlation."""
    return str(uuid.uuid4())[:8]


def _handle_error(
    response: httpx.Response,
    agent_name: str | None = None,
    bot_name: str | None = None,
    request_id: str | None = None,
) -> None:
    """
    Handle HTTP error responses.

    Args:
        response: The HTTP response
        agent_name: Agent name for error context
        bot_name: Bot name for error context
        request_id: Request ID for correlation

    Raises:
        AuthenticationError: For 401 responses
        AgentNotEnabledError: For 403 responses with agent context
        AgentNotFoundError: For 404 responses with agent context
        BotNotFoundError: For 404 responses with bot context
        RateLimitError: For 429 responses
        AgentExecutionError: For other 4xx/5xx responses
    """
    context = f"[req:{request_id}]" if request_id else ""

    if response.status_code == 401:
        logger.warning("%s Authentication failed", context)
        raise AuthenticationError()

    if response.status_code == 403:
        if agent_name:
            logger.warning("%s Agent not enabled: %s", context, agent_name)
            raise AgentNotEnabledError(agent_name)
        raise MetadataError("Access forbidden", status_code=403)

    if response.status_code == 404:
        if agent_name:
            logger.warning("%s Agent not found: %s", context, agent_name)
            raise AgentNotFoundError(agent_name)
        if bot_name:
            logger.warning("%s Bot not found: %s", context, bot_name)
            raise BotNotFoundError(bot_name)
        raise MetadataError("Resource not found", status_code=404)

    if response.status_code == 429:
        retry_after = response.headers.get("Retry-After")
        retry_seconds = int(retry_after) if retry_after else None
        logger.warning("%s Rate limit exceeded, retry after: %s", context, retry_seconds)
        raise RateLimitError("Rate limit exceeded", retry_after=retry_seconds)

    if response.status_code >= 400:
        # LBYL: Check content-type before parsing JSON
        content_type = response.headers.get("content-type", "")
        if "application/json" in content_type:
            error_data = response.json()
            message = error_data.get("message", response.text)
        else:
            message = response.text

        logger.error(
            "%s API error (%d): %s",
            context,
            response.status_code,
            message[:200],
        )
        raise AgentExecutionError(
            f"API error ({response.status_code}): {message}",
            agent_name=agent_name,
        )


class HTTPClient:
    """
    HTTP client for API communication with retry support.

    Features:
    - Automatic retry with exponential backoff
    - Request correlation IDs
    - Proper logging
    - Connection pooling via httpx
    """

    def __init__(
        self,
        base_url: str,
        auth: TokenAuth,
        timeout: float = 120.0,
        verify_ssl: bool = True,
        max_retries: int = 3,
        retry_delay: float = 1.0,
        user_agent: str | None = None,
    ):
        """
        Initialize the HTTP client.

        Args:
            base_url: Base URL for API requests
            auth: Authentication handler
            timeout: Request timeout in seconds
            verify_ssl: Whether to verify SSL certificates
            max_retries: Maximum number of retry attempts
            retry_delay: Base delay between retries (exponential backoff)
            user_agent: Custom User-Agent string
        """
        self._base_url = base_url.rstrip("/")
        self._auth = auth
        self._timeout = timeout
        self._verify_ssl = verify_ssl
        self._max_retries = max_retries
        self._retry_delay = retry_delay
        self._user_agent = user_agent or "metadata-ai-sdk/0.1.0"

        self._client = httpx.Client(
            base_url=self._base_url,
            timeout=timeout,
            verify=verify_ssl,
        )

        logger.debug("HTTPClient initialized for %s", self._base_url)

    def _headers(self, request_id: str | None = None) -> dict[str, str]:
        """Get request headers."""
        headers = self._auth.get_headers()
        headers["Content-Type"] = "application/json"
        headers["Accept"] = "application/json"
        headers["User-Agent"] = self._user_agent
        if request_id:
            headers["X-Request-ID"] = request_id
        return headers

    def _should_retry(self, response: httpx.Response, attempt: int) -> bool:
        """Check if request should be retried."""
        if attempt >= self._max_retries:
            return False
        return response.status_code in RETRYABLE_STATUS_CODES

    def _wait_for_retry(self, attempt: int, response: httpx.Response) -> None:
        """Wait before retrying with exponential backoff."""
        # Check for Retry-After header
        retry_after = response.headers.get("Retry-After")
        if retry_after:
            # LBYL: Check if it's a valid number before converting
            if retry_after.replace(".", "", 1).isdigit():
                delay = float(retry_after)
            else:
                delay = self._retry_delay * (2**attempt)
        else:
            delay = self._retry_delay * (2**attempt)

        logger.debug("Retry attempt %d, waiting %.2fs", attempt + 1, delay)
        time.sleep(delay)

    def get(
        self,
        path: str,
        params: dict[str, Any] | None = None,
        agent_name: str | None = None,
        bot_name: str | None = None,
    ) -> dict[str, Any]:
        """
        Make a GET request with retry support.

        Args:
            path: Request path
            params: Query parameters
            agent_name: Agent name for error context
            bot_name: Bot name for error context

        Returns:
            Response JSON data
        """
        request_id = _generate_request_id()
        logger.debug("[req:%s] GET %s%s", request_id, self._base_url, path)

        last_response = None
        for attempt in range(self._max_retries + 1):
            response = self._client.get(
                path,
                headers=self._headers(request_id),
                params=params,
            )
            last_response = response

            if response.status_code < 400:
                return response.json()

            if self._should_retry(response, attempt):
                self._wait_for_retry(attempt, response)
                continue

            break

        assert last_response is not None  # Loop always runs at least once
        _handle_error(last_response, agent_name=agent_name, bot_name=bot_name, request_id=request_id)
        return {}  # Never reached, _handle_error always raises

    def post(
        self,
        path: str,
        json: dict[str, Any],
        agent_name: str | None = None,
    ) -> dict[str, Any]:
        """
        Make a POST request with retry support.

        Args:
            path: Request path
            json: Request body
            agent_name: Agent name for error context

        Returns:
            Response JSON data
        """
        request_id = _generate_request_id()
        logger.debug("[req:%s] POST %s%s", request_id, self._base_url, path)
        logger.debug("[req:%s] Request body: %s", request_id, json_module.dumps(json, indent=2))

        last_response = None
        for attempt in range(self._max_retries + 1):
            response = self._client.post(
                path,
                headers=self._headers(request_id),
                json=json,
            )
            last_response = response

            if response.status_code < 400:
                result = response.json()
                logger.debug(
                    "[req:%s] Response status: %d",
                    request_id,
                    response.status_code,
                )
                return result

            if self._should_retry(response, attempt):
                self._wait_for_retry(attempt, response)
                continue

            break

        assert last_response is not None  # Loop always runs at least once
        _handle_error(last_response, agent_name=agent_name, request_id=request_id)
        return {}  # Never reached

    def post_stream(
        self,
        path: str,
        json: dict[str, Any],
        agent_name: str | None = None,
    ) -> Iterator[bytes]:
        """
        Make a streaming POST request.

        Note: Streaming requests don't support automatic retry.

        Args:
            path: Request path
            json: Request body
            agent_name: Agent name for error context

        Yields:
            Response chunks as bytes
        """
        request_id = _generate_request_id()
        logger.debug("[req:%s] POST (stream) %s%s", request_id, self._base_url, path)

        headers = self._auth.get_headers()
        headers["Content-Type"] = "application/json"
        headers["Accept"] = "text/event-stream"
        headers["User-Agent"] = self._user_agent
        headers["X-Request-ID"] = request_id

        with self._client.stream(
            "POST",
            path,
            headers=headers,
            json=json,
        ) as response:
            # Check for errors before streaming
            if response.status_code >= 400:
                response.read()
                _handle_error(response, agent_name=agent_name, request_id=request_id)

            yield from response.iter_bytes()

    def close(self) -> None:
        """Close the HTTP client."""
        logger.debug("Closing HTTPClient")
        self._client.close()

    def __enter__(self) -> HTTPClient:
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()


class AsyncHTTPClient:
    """
    Async HTTP client for API communication with retry support.

    Features:
    - Automatic retry with exponential backoff
    - Request correlation IDs
    - Proper logging
    - Lazy client initialization
    """

    def __init__(
        self,
        base_url: str,
        auth: TokenAuth,
        timeout: float = 120.0,
        verify_ssl: bool = True,
        max_retries: int = 3,
        retry_delay: float = 1.0,
        user_agent: str | None = None,
    ):
        """
        Initialize the async HTTP client.

        Args:
            base_url: Base URL for API requests
            auth: Authentication handler
            timeout: Request timeout in seconds
            verify_ssl: Whether to verify SSL certificates
            max_retries: Maximum number of retry attempts
            retry_delay: Base delay between retries
            user_agent: Custom User-Agent string
        """
        self._base_url = base_url.rstrip("/")
        self._auth = auth
        self._timeout = timeout
        self._verify_ssl = verify_ssl
        self._max_retries = max_retries
        self._retry_delay = retry_delay
        self._user_agent = user_agent or "metadata-ai-sdk/0.1.0"
        self._client: httpx.AsyncClient | None = None

        logger.debug("AsyncHTTPClient initialized for %s", self._base_url)

    def _get_client(self) -> httpx.AsyncClient:
        """Get or create the async client."""
        if self._client is None:
            self._client = httpx.AsyncClient(
                base_url=self._base_url,
                timeout=self._timeout,
                verify=self._verify_ssl,
            )
        return self._client

    def _headers(self, request_id: str | None = None) -> dict[str, str]:
        """Get request headers."""
        headers = self._auth.get_headers()
        headers["Content-Type"] = "application/json"
        headers["Accept"] = "application/json"
        headers["User-Agent"] = self._user_agent
        if request_id:
            headers["X-Request-ID"] = request_id
        return headers

    def _should_retry(self, response: httpx.Response, attempt: int) -> bool:
        """Check if request should be retried."""
        if attempt >= self._max_retries:
            return False
        return response.status_code in RETRYABLE_STATUS_CODES

    async def _wait_for_retry(self, attempt: int, response: httpx.Response) -> None:
        """Wait before retrying with exponential backoff."""
        import asyncio

        retry_after = response.headers.get("Retry-After")
        if retry_after:
            # LBYL: Check if it's a valid number before converting
            if retry_after.replace(".", "", 1).isdigit():
                delay = float(retry_after)
            else:
                delay = self._retry_delay * (2**attempt)
        else:
            delay = self._retry_delay * (2**attempt)

        logger.debug("Async retry attempt %d, waiting %.2fs", attempt + 1, delay)
        await asyncio.sleep(delay)

    async def get(
        self,
        path: str,
        params: dict[str, Any] | None = None,
        agent_name: str | None = None,
        bot_name: str | None = None,
    ) -> dict[str, Any]:
        """
        Make an async GET request with retry support.

        Args:
            path: Request path
            params: Query parameters
            agent_name: Agent name for error context
            bot_name: Bot name for error context

        Returns:
            Response JSON data
        """
        request_id = _generate_request_id()
        logger.debug("[req:%s] async GET %s%s", request_id, self._base_url, path)

        client = self._get_client()
        last_response = None

        for attempt in range(self._max_retries + 1):
            response = await client.get(
                path,
                headers=self._headers(request_id),
                params=params,
            )
            last_response = response

            if response.status_code < 400:
                return response.json()

            if self._should_retry(response, attempt):
                await self._wait_for_retry(attempt, response)
                continue

            break

        assert last_response is not None  # Loop always runs at least once
        _handle_error(last_response, agent_name=agent_name, bot_name=bot_name, request_id=request_id)
        return {}  # Never reached

    async def post(
        self,
        path: str,
        json: dict[str, Any],
        agent_name: str | None = None,
    ) -> dict[str, Any]:
        """
        Make an async POST request with retry support.

        Args:
            path: Request path
            json: Request body
            agent_name: Agent name for error context

        Returns:
            Response JSON data
        """
        request_id = _generate_request_id()
        logger.debug("[req:%s] async POST %s%s", request_id, self._base_url, path)

        client = self._get_client()
        last_response = None

        for attempt in range(self._max_retries + 1):
            response = await client.post(
                path,
                headers=self._headers(request_id),
                json=json,
            )
            last_response = response

            if response.status_code < 400:
                return response.json()

            if self._should_retry(response, attempt):
                await self._wait_for_retry(attempt, response)
                continue

            break

        assert last_response is not None  # Loop always runs at least once
        _handle_error(last_response, agent_name=agent_name, request_id=request_id)
        return {}  # Never reached

    async def post_stream(
        self,
        path: str,
        json: dict[str, Any],
        agent_name: str | None = None,
    ) -> AsyncIterator[bytes]:
        """
        Make an async streaming POST request.

        Note: Streaming requests don't support automatic retry.

        Args:
            path: Request path
            json: Request body
            agent_name: Agent name for error context

        Yields:
            Response chunks as bytes
        """
        request_id = _generate_request_id()
        logger.debug("[req:%s] async POST (stream) %s%s", request_id, self._base_url, path)

        headers = self._auth.get_headers()
        headers["Content-Type"] = "application/json"
        headers["Accept"] = "text/event-stream"
        headers["User-Agent"] = self._user_agent
        headers["X-Request-ID"] = request_id

        client = self._get_client()
        async with client.stream(
            "POST",
            path,
            headers=headers,
            json=json,
        ) as response:
            # Check for errors before streaming
            if response.status_code >= 400:
                await response.aread()
                _handle_error(response, agent_name=agent_name, request_id=request_id)

            async for chunk in response.aiter_bytes():
                yield chunk

    async def close(self) -> None:
        """Close the async HTTP client."""
        logger.debug("Closing AsyncHTTPClient")
        if self._client is not None:
            await self._client.aclose()
            self._client = None

    async def __aenter__(self) -> AsyncHTTPClient:
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.close()

File: typescript/src/http.ts
/**
 * HTTP client for the Metadata AI SDK.
 *
 * This module provides a thin wrapper around the native fetch API
 * with support for:
 * - Automatic retry with exponential backoff
 * - Request correlation IDs for debugging
 * - Proper error handling and mapping
 * - Custom base URL support for different API endpoints
 */

import {
  AbilityNotFoundError,
  AgentExecutionError,
  AgentNotEnabledError,
  AgentNotFoundError,
  AuthenticationError,
  BotNotFoundError,
  AISdkError,
  NetworkError,
  PersonaNotFoundError,
  RateLimitError,
  TimeoutError,
} from './errors.js';

/** Status codes that should trigger a retry */
const RETRYABLE_STATUS_CODES = new Set([429, 500, 502, 503, 504]);

/**
 * Generate a unique request ID for correlation.
 */
function generateRequestId(): string {
  return Math.random().toString(36).substring(2, 10);
}

/**
 * Options for HTTP client configuration.
 */
export interface HttpClientOptions {
  /** Base URL for API requests */
  baseUrl: string;
  /** JWT token for authentication */
  token: string;
  /** Request timeout in milliseconds */
  timeout: number;
  /** Maximum number of retry attempts */
  maxRetries: number;
  /** Base delay between retries in milliseconds */
  retryDelay: number;
}

/**
 * Entity type for error context.
 */
export type EntityType = 'agent' | 'bot' | 'persona' | 'ability';

/**
 * Options for individual HTTP requests.
 */
export interface RequestOptions {
  /** HTTP method */
  method: 'GET' | 'POST' | 'DELETE';
  /** Request path (appended to baseUrl) */
  path: string;
  /** Request body (for POST requests) */
  body?: unknown;
  /** Query parameters (for GET/DELETE requests) */
  params?: Record<string, string | number | boolean>;
  /** Agent name for error context */
  agentName?: string;
  /** Entity type for error context (default: 'agent') */
  entityType?: EntityType;
  /** Entity name for error context */
  entityName?: string;
  /** Whether this is a streaming request */
  stream?: boolean;
}

/**
 * HTTP client for API communication with retry support.
 */
export class HttpClient {
  private readonly baseUrl: string;
  private readonly token: string;
  private readonly timeout: number;
  private readonly maxRetries: number;
  private readonly retryDelay: number;
  private readonly userAgent = 'ai-sdk-ts/0.0.2';

  constructor(options: HttpClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, '');
    this.token = options.token;
    this.timeout = options.timeout;
    this.maxRetries = options.maxRetries;
    this.retryDelay = options.retryDelay;
  }

  /**
   * Build request headers.
   */
  private getHeaders(requestId: string, stream = false): Record<string, string> {
    return {
      Authorization: `Bearer ${this.token}`,
      'Content-Type': 'application/json',
      Accept: stream ? 'text/event-stream' : 'application/json',
      'User-Agent': this.userAgent,
      'X-Request-ID': requestId,
    };
  }

  /**
   * Build full URL with query parameters.
   */
  private buildUrl(
    path: string,
    params?: Record<string, string | number | boolean>
  ): string {
    // Construct full URL by appending path to base URL
    const fullPath = path.startsWith('/') ? path : `/${path}`;
    const urlString = `${this.baseUrl}${fullPath}`;
    const url = new URL(urlString);
    if (params) {
      for (const [key, value] of Object.entries(params)) {
        url.searchParams.set(key, String(value));
      }
    }
    return url.toString();
  }

  /**
   * Check if a response status should trigger a retry.
   */
  private shouldRetry(status: number, attempt: number): boolean {
    if (attempt >= this.maxRetries) {
      return false;
    }
    return RETRYABLE_STATUS_CODES.has(status);
  }

  /**
   * Calculate retry delay with exponential backoff.
   */
  private getRetryDelay(attempt: number, retryAfterHeader?: string | null): number {
    if (retryAfterHeader) {
      const parsed = parseInt(retryAfterHeader, 10);
      if (!isNaN(parsed)) {
        return parsed * 1000; // Convert seconds to milliseconds
      }
    }
    return this.retryDelay * Math.pow(2, attempt);
  }

  /**
   * Wait for the specified delay.
   */
  private async wait(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  /**
   * Handle error responses and throw appropriate exceptions.
   */
  private async handleError(
    response: Response,
    agentName?: string,
    _requestId?: string,
    entityType?: EntityType,
    entityName?: string
  ): Promise<never> {
    const status = response.status;

    if (status === 401) {
      throw new AuthenticationError();
    }

    if (status === 403) {
      if (agentName) {
        throw new AgentNotEnabledError(agentName);
      }
      throw new AISdkError('Access forbidden', 403);
    }

    if (status === 404) {
      // Check entity type for specific error classes
      if (entityType === 'bot' && entityName) {
        throw new BotNotFoundError(entityName);
      }
      if (entityType === 'persona' && entityName) {
        throw new PersonaNotFoundError(entityName);
      }
      if (entityType === 'ability' && entityName) {
        throw new AbilityNotFoundError(entityName);
      }
      if (agentName) {
        throw new AgentNotFoundError(agentName);
      }
      throw new AISdkError('Resource not found', 404);
    }

    if (status === 429) {
      const retryAfter = response.headers.get('Retry-After');
      const retrySeconds = retryAfter ? parseInt(retryAfter, 10) : undefined;
      throw new RateLimitError('Rate limit exceeded', retrySeconds);
    }

    // Try to extract error message from response body
    let message: string;
    const contentType = response.headers.get('content-type') || '';
    if (contentType.includes('application/json')) {
      try {
        const errorData = (await response.json()) as { message?: string };
        message = errorData.message || response.statusText;
      } catch {
        message = response.statusText;
      }
    } else {
      message = await response.text().catch(() => response.statusText);
    }

    throw new AgentExecutionError(
      `API error (${status}): ${message}`,
      agentName
    );
  }

  /**
   * Make a GET request with retry support.
   */
  async get<T>(
    path: string,
    params?: Record<string, string | number | boolean>,
    agentName?: string,
    entityType?: EntityType,
    entityName?: string
  ): Promise<T> {
    return this.request<T>({
      method: 'GET',
      path,
      params,
      agentName,
      entityType,
      entityName,
    });
  }

  /**
   * Make a POST request with retry support.
   */
  async post<T>(
    path: string,
    body: unknown,
    agentName?: string,
    entityType?: EntityType,
    entityName?: string
  ): Promise<T> {
    return this.request<T>({
      method: 'POST',
      path,
      body,
      agentName,
      entityType,
      entityName,
    });
  }

  /**
   * Make a DELETE request with retry support.
   *
   * Returns void for empty responses; otherwise the parsed JSON body.
   */
  async delete<T = void>(
    path: string,
    params?: Record<string, string | number | boolean>,
    agentName?: string,
    entityType?: EntityType,
    entityName?: string
  ): Promise<T> {
    return this.request<T>({
      method: 'DELETE',
      path,
      params,
      agentName,
      entityType,
      entityName,
    });
  }

  /**
   * Make a streaming POST request.
   *
   * Note: Streaming requests don't support automatic retry.
   *
   * Unlike non-streaming methods, `postStream` intentionally does NOT apply
   * `this.timeout` as a request deadline. SSE streams can run for many
   * minutes (long agent runs with multiple tool calls), and the stream's
   * own events signal liveness. A fixed deadline would cut the stream
   * mid-run regardless of progress. Callers that need cancellation can
   * abort externally (e.g. close the underlying ReadableStream / Ctrl+C).
   */
  async postStream(
    path: string,
    body: unknown,
    agentName?: string
  ): Promise<ReadableStream<Uint8Array>> {
    const requestId = generateRequestId();
    const url = this.buildUrl(path);
    const headers = this.getHeaders(requestId, true);

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers,
        body: JSON.stringify(body),
      });

      if (!response.ok) {
        await this.handleError(response, agentName, requestId);
      }

      if (!response.body) {
        throw new AISdkError('No response body for streaming request');
      }

      return response.body;
    } catch (error) {
      if (error instanceof AISdkError) {
        throw error;
      }

      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          // Caller-initiated abort (no internal deadline is applied here).
          throw new TimeoutError(this.timeout);
        }
        throw new NetworkError(`Network error: ${error.message}`, error);
      }

      throw new NetworkError('Unknown network error');
    }
  }

  /**
   * Parse the response body. Returns undefined for empty / 204 responses
   * (e.g. DELETE) so callers can declare a `void` return type.
   */
  private async parseBody<T>(response: Response): Promise<T> {
    if (response.status === 204) {
      return undefined as T;
    }
    const contentLength = response.headers?.get?.('content-length');
    if (contentLength === '0') {
      return undefined as T;
    }
    try {
      return (await response.json()) as T;
    } catch {
      return undefined as T;
    }
  }

  /**
   * Internal request method with retry logic.
   */
  private async request<T>(options: RequestOptions): Promise<T> {
    const { method, path, body, params, agentName, entityType, entityName } = options;
    const requestId = generateRequestId();
    const url = this.buildUrl(path, params);
    const headers = this.getHeaders(requestId);

    let lastError: Error | null = null;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);

      try {
        const response = await fetch(url, {
          method,
          headers,
          body: body ? JSON.stringify(body) : undefined,
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        if (response.ok) {
          return await this.parseBody<T>(response);
        }

        // Check if we should retry
        if (this.shouldRetry(response.status, attempt)) {
          const delay = this.getRetryDelay(
            attempt,
            response.headers.get('Retry-After')
          );
          await this.wait(delay);
          continue;
        }

        // Non-retryable error
        await this.handleError(response, agentName, requestId, entityType, entityName);
      } catch (error) {
        clearTimeout(timeoutId);

        if (error instanceof AISdkError) {
          throw error;
        }

        if (error instanceof Error) {
          if (error.name === 'AbortError') {
            lastError = new TimeoutError(this.timeout);
            // Timeouts are retryable
            if (attempt < this.maxRetries) {
              const delay = this.retryDelay * Math.pow(2, attempt);
              await this.wait(delay);
              continue;
            }
            throw lastError;
          }
          lastError = new NetworkError(`Network error: ${error.message}`, error);
          // Network errors are retryable
          if (attempt < this.maxRetries) {
            const delay = this.retryDelay * Math.pow(2, attempt);
            await this.wait(delay);
            continue;
          }
          throw lastError;
        }

        throw new NetworkError('Unknown network error');
      }
    }

    // Should never reach here, but TypeScript needs this
    throw lastError || new NetworkError('Request failed after retries');
  }
}
